ably/realtime_channel.go

```diff
@@ -462,24 +428,65 @@ func (c *RealtimeChannel) History(params *PaginateParams) (*PaginatedResult, err
 }
 
 func (c *RealtimeChannel) send(msg *proto.ProtocolMessage) (Result, error) {
-	if _, err := c.attach(false); err != nil {
-		return nil, err
-	}
-	res, listen := newErrResult()
-	switch c.State() {
-	case ChannelStateInitialized, ChannelStateAttaching:
-		c.queue.Enqueue(msg, listen)
+	if res, enqueued := c.maybeEnqueue(msg); enqueued {
```

Comment:
Given `Publish` is now blocking but takes a context, should this queue not take into account the context deadline? For example, if a connection is stuck in a connecting / disconnecting loop and I publish with a 5s timeout, I'd probably expect that to return with an error after 5s?

Reply:
The contexts (here and elsewhere) are handled at the outer layer. The pattern is:

* Public function takes a context and calls an internal function.
* Internal function starts some process and returns a Result to wait for it. (Here, the Result waits for the message to be dequeued, and either be failed or published and then acknowledged.)
* Public function waits for the Result, but will stop waiting if the context is called.

This implements [what we decided](https://github.com/ably/ably-go/issues/197#issuecomment-696979466) in terms of what a context means in the Realtime API. (In REST, contexts will be passed all the way to the HTTP library.)

Reply:
Ah I'm sorry, I thought that `Publish` would not return when the context was cancelled, but I see I was incorrect 👍